### PR TITLE
Loading assets from disk

### PR DESCRIPTION
**Pull request by vswarte**
Messy draft PR to track work on the file-loading hook.

The end goal is to allow asset overrides like modengine2 allows while supporting the same folder structure such that we can load existing mods. Unfortunately the hook(s) used to achieve this in modengine2 is kinda smelly and ideally dropped for something a bit easier to maintain.

This PR implements a similar feature by hooking RSResourceFileRequest::vtable[1] (RSResourceFileRequest::IsPathStringEmpty unofficial name). Which seems to be called before actually handling file requests, granting us consistent access to the files path before the actual IO system does its thing.

In order to operate on the DLWstring that hosts the file path @Dasaav-dsv has modeled DLWString in C++ using `std::basic_string`. This PR introduces cxx as a dependency and a bunch of C++ code for that reason.

An important difference between me2 and me3's approach is that me2 was receiving fully expanded paths mapped to `dataX` and the like directly. This hook unfortunately gets us unexpanded paths eg: `event:/common.emevd.dcx`. This can be solved by asking the game to resolve it for us, acquiring the mapping from the games memory ourselves or keeping a manual table per game.

What still needs to happen:
- [x] Look into better ways to attach extensions like this to mod host
  - [x] Remove the need for making the package source `pub`
- [x] Virtual root mapping
  - [x] Look into possibilities with fetching it from the game dynamically?
  - [x] Fixed mapping with additional manual work as plan B?
- [x] Dynamic hook locating
  - [x] Vftable crawler for ER
  - [ ] AOB for AC6
- [x] Wwise hook

To cap it all off, I couldn't immediately get the code to run with the dependency and package ordering code on, still need to look into that 😄

### CODE REVIEW

File: crates/mod-host-assets/include/dl_allocator.hpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#ifndef DANTELION_DL_ALLOCATOR_H
```
Comment (garyttierney) [inline crates/mod-host-assets/include/dl_allocator.hpp:1]: `#pragma once` for include guards, we don't care about standard C++ here.

File: crates/mod-host-assets/include/dl_allocator.hpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    template <typename T1, typename T2>
+    friend bool operator==(const DLAllocatorAdapter<T1>& lhs, const DLAllocatorAdapter<T2>& rhs) noexcept;
+
+private:
+    DLAllocationInterface& allocator;
+};
+
+template <typename T1, typename T2>
```
Comment (garyttierney) [inline crates/mod-host-assets/include/dl_allocator.hpp:54]: Why are the definition and declaration separated here?

File: crates/mod-host-assets/include/dl_allocator.hpp (hunk 1 above)
Comment (vswarte) [reply crates/mod-host-assets/include/dl_allocator.hpp]: https://github.com/garyttierney/me3/pull/22/commits/3724dbf6abbd846c3750f251d1be7257556e15ab

File: crates/mod-host-assets/include/dl_allocator.hpp (hunk 2 above)
Comment (vswarte) [reply crates/mod-host-assets/include/dl_allocator.hpp]: https://github.com/garyttierney/me3/pull/22/commits/ba5286e18c81731ad078e74f22fb406af00ea148

File: crates/launcher/src/main.rs [hunk 3]
```diff
@@ -54,14 +54,16 @@ fn run() -> LauncherResult<()> {
 
     // TODO: merge
     if let Some(mut profile) = profiles.into_iter().next() {
-        let ordered_natives = sort_dependencies(profile.natives())
```
Comment (garyttierney) [inline crates/launcher/src/main.rs:57]: Can you share the config this code is causing problems with?

Comment (vswarte) [reply crates/launcher/src/main.rs:57]:
```toml
profileVersion = "v1"

[[natives]]
path = "waygate_client.dll"
optional = true

[[natives]]
path = "eldenring_alt_saves.dll"
optional = true

# [[packages]]
# id = "test-mod-emevd-rave"
# source = "Z:\\home\\vincent\\me3-test-mod\\emevd-rave"
#
# [[packages]]
# id = "test-mod-clevers"
# source = "Z:\\home\\vincent\\me3-test-mod\\clevers"

[[packages]]
id = "err"
source = "Z:\\home\\vincent\\me3-test-mod\\err"
```

**Comment by vswarte** [comment]
The new hook had some nasty edge-cases around finalizing map loads causing the user to indefinitely be stuck. I'm updating the code to use the old hook again but since the old hook is taking in a `DLWString` (allocator included) we no longer need the second `CreateFileW` hook.

File: crates/mod-host/src/asset_archive.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+                // Match the expanded path against the known overrides.
+                if let Some(mapped_override) = mapping.get_override(&resource_path_string) {
+                    // Replace the string with a canonical path to the asset if
+                    // we did find an override. This will cause the game to 
+                    // pull the files bytes from the file system instead of the
+                    // BDTs.
+                    set_dlwstring_contents(unsafe { path.as_ref().unwrap() }, mapped_override);
+
+                    writeln!(
```
Comment (garyttierney) [inline crates/mod-host/src/asset_archive.rs:50]: Should use `tracing` crate log functions.

Comment (vswarte) [reply crates/mod-host/src/asset_archive.rs]: Fixed in [57c7bf](https://github.com/garyttierney/me3/pull/22/commits/157c7bfffffbf886bb184d7cc915aff16e1daf14)

**Review by garyttierney — APPROVED** [review]

**Comment by vswarte** [comment]
Rebased this on branch that fixes the build

**Comment by vswarte** [comment]
The last remaining item for this I think is resolving relative paths. With the current structure of `AttachRequest` the mod host does not know the path of the mod profile declaring some package or native. This means that we cannot resolve relative paths either.

**Comment by vswarte** [comment]
We discussed the proper approach and it seems the idea is to make the path absolute before sending it as part of the `AttachRequest`. This means that the host will get absolute paths and the launcher is responsible for resolving them.
